proto/src/op/message.rs

```diff
@@ -633,7 +633,7 @@ impl Message {
     ///
     /// Subsequent to calling this, the Message should not change.
     #[cfg(feature = "openssl")]
-    pub fn sign(&mut self, signer: &Signer, inception_time: u32) -> DnsSecResult<()> {
+    pub fn sign<S: MessageSigner>(&mut self, signer: &S, inception_time: u32) -> ProtoResult<()> {
```

Comment:
@briansmith I still have a way to go, but you can see the direction I'm going here. MessageSigner will be a generic type defined in the proto crate, with implementations in the Client crate.

Reply:
AFAICT, you don't need to parameterize `sign()` by `<S: MessageSigner>`. You can just do `signer: &MessageSigner`. Simpler and less code bloat from monopolization.

Reply:
I guess you can replace `[cfg(feature = "openssl")]` either with `#[cfg(any(feature = "openssl", feature = "ring"))]` or by making it available unconditionally.

Reply:
I think the code for this entire function can/should be kept in `client`, ideally. That could happen later.

Reply:
`&MessageSigner` is definitely simpler, I was trying to avoid a vtable lookup since I believe people will be using one or the other. To make this change easier, I might to a trait object for now, and then revisit.

>this entire function can/should be kept in `client`

I might change the name of it and move some of the functionality around. The problem is that I want this to happen after some other standard ops, like `query_id` generation. To do that it needs to happen just prior to sending the packet.

An option I'm toying with is allowing for a generic type to be associated to the new DnsHandle which would perform some final evaluation of the message before sending.